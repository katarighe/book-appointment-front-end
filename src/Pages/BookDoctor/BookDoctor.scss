.bookDoctor {
  //   background: var(--mainColor);

  margin: 2em auto;
}
