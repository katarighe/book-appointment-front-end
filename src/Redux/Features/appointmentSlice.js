import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import * as API from '../../api/apis';

const initialState = {
  allAppointments: [],
  isLoading: false,
  isError: '',
  appointmentDetails: null,
};

export const createAppointment = createAsyncThunk(
  'appointments/createAppointment',
  async (userData, thunkAPI) => {
    try {
      const response = await API.createAppointment(userData);
      return response.data.data;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.response.data);
    }
  },
);

export const getAppointments = createAsyncThunk(
  'appointments/getAppointments',
  async (_, thunkAPI) => {
    try {
      const response = await API.getUserAppointments();
      return response.data.data;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.response.data);
    }
  },
);

export const appointmentSlice = createSlice({
  name: 'appointments',
  initialState,

  reducers: {
    getAppointmentDetails: (state, action) => {
      state.appointmentDetails = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(createAppointment.pending, (state) => {
        state.isLoading = true;
      })
      .addCase(createAppointment.fulfilled, (state, action) => {
        state.isLoading = false;
        state.isError = '';
        state.allAppointments.push(action.payload);
      })
      .addCase(createAppointment.rejected, (state, action) => {
        state.isLoading = false;
        state.isError = action.payload;
      })
      .addCase(getAppointments.pending, (state) => {
        state.isLoading = true;
      })
      .addCase(getAppointments.fulfilled, (state, action) => {
        state.isLoading = false;
        state.isError = '';
        state.allAppointments = action.payload;
      })
      .addCase(getAppointments.rejected, (state, action) => {
        state.isLoading = false;
        state.isError = action.payload
          ? action.payload.error
          : 'An error occurred';
      });
  },
});

export const { getAppointmentDetails } = appointmentSlice.actions;

export const selectAppointment = (state) => state.appointmentSlice;

export default appointmentSlice.reducer;
