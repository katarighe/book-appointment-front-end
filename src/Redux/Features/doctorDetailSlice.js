import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios'

const url = '';

export const fetchDoctorDetails = createAsyncThunk('fetchDoctorDetails', async () => {
  const response = await axios.get('http://127.0.0.1:3000/url', {headers: {
    Accept: 'application/json',
  }});
  return response.data
});

const initialState = {
    value: {},
    loading: false,
    error : '',
  }

  const DoctorDetailSlice = createSlice({
    name: 'DoctorDetails',
    initialState,
    reducers: {},
    extraReducers: (builder) => {
      builder
        .addCase(fetchDoctorDetails.fulfilled, (state, action) => ({...state, value: action.payload}));
    },
  });

  export default DoctorDetailSlice.reducers;
  