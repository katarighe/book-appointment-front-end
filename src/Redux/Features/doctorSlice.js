import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import * as API from '../../api/apis';

const initialState = {
  allDoctors: [],
  isLoading: false,
  isError: '',
};

export const addDoctors = createAsyncThunk(
  'doctors/addDoctors',
  async (userData, thunkAPI) => {
    try {
      const response = await API.addDoctors(userData);
      return response.data.data;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.response.data);
    }
  },
);

export const deleteDoctors = createAsyncThunk(
  'doctors/deleteDoctors',
  async (id, thunkAPI) => {
    try {
      const response = await API.deleteDoctors(id);

      return response;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.response.data);
    }
  },
);

export const getDoctors = createAsyncThunk(
  'doctors/getDoctors',
  async (_, thunkAPI) => {
    try {
      const response = await API.getDoctors();
      return response.data.data;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.response.data);
    }
  },
);

export const doctorSlice = createSlice({
  name: 'doctors',
  initialState,

  reducers: {
    removeDoctor: (state, action) => {
      const docId = action.payload;
      state.allDoctors = state.allDoctors.filter((doc) => doc.id !== docId);
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(addDoctors.pending, (state) => {
        state.isLoading = true;
      })
      .addCase(addDoctors.fulfilled, (state, action) => {
        state.isLoading = false;
        state.isError = '';
        state.allDoctors.push(action.payload);
      })
      .addCase(addDoctors.rejected, (state, action) => {
        state.isLoading = false;
        state.isError = action.payload;
      })
      .addCase(getDoctors.pending, (state) => {
        state.isLoading = true;
      })
      .addCase(getDoctors.fulfilled, (state, action) => {
        state.isLoading = false;
        state.isError = '';
        state.allDoctors = action.payload;
      })
      .addCase(getDoctors.rejected, (state, action) => {
        state.isLoading = false;
        state.isError = action.payload
          ? action.payload.error
          : 'An error occurred';
      })
      .addCase(deleteDoctors.pending, (state) => {
        state.isLoading = true;
      })
      .addCase(deleteDoctors.fulfilled, (state) => {
        state.isLoading = false;
        state.isError = '';
      })
      .addCase(deleteDoctors.rejected, (state, action) => {
        state.isLoading = false;
        state.isError = action.payload
          ? action.payload.error
          : 'An error occurred';
      });
  },
});

export const { removeDoctor } = doctorSlice.actions;
export const selectDoctor = (state) => state.doctorSlice;
export default doctorSlice.reducer;
