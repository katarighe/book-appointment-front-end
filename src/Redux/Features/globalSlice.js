import { createSlice } from '@reduxjs/toolkit';

export const globalSlice = createSlice({
  name: 'global',
  initialState: {
    count: 0,
  },

  reducers: {
    toggleShow: (state, action) => {
      // Toggle icon state based on the IDs
      const id = action.payload;
      return {
        ...state,
        [id]: !state[id],
      };
    },

    increment: (state, action) => {
      state.count = action.payload;
    },
  },
});

export const { toggleShow, increment } = globalSlice.actions;

export const selectGlobal = (state) => state.globalSlice;
export default globalSlice.reducer;
