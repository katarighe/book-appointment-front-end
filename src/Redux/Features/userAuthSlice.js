import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  authUser: {},
  isLoggedIn: false,
};

export const userAuthSlice = createSlice({
  name: 'userDatas',
  initialState,

  reducers: {
    userAuthData: (state, action) => {
      state.authUser = action.payload;
      state.isLoggedIn = true;
    },

    userLogOut: (state) => {
      state.authUser = null;
      state.isLoggedIn = false;
    },
  },
});

export const { userAuthData, userLogOut } = userAuthSlice.actions;

export const selectUserData = (state) => state.userAuthSlice;

export default userAuthSlice.reducer;
