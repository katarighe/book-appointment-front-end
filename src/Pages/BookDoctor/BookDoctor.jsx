import { useState } from 'react';
import { useSweetAlert } from '../../Hooks/useSweetAlert';
import './BookDoctor.scss';
import { useNavigate } from 'react-router-dom';
import { BsCaretLeft } from 'react-icons/bs';
import { useDispatch, useSelector } from 'react-redux';
import { selectUserData } from '../../Redux/Features/userAuthSlice';
import {
  createAppointment,
  selectAppointment,
} from '../../Redux/Features/appointmentSlice';
import { Spinner } from 'react-bootstrap';

const BookDoctor = () => {
  const { showAlert } = useSweetAlert();
  const dispatch = useDispatch();
  const { appointmentDetails, isLoading } = useSelector(selectAppointment);
  const { authUser } = useSelector(selectUserData);

  const [bookingData, setBookingData] = useState({
    date_of_appointment: '',
    user_id: authUser.image?.record.id,
    doctor_id: appointmentDetails.doctor_id,
    username: authUser.name,
    doctorname: appointmentDetails.name,
  });

  // Function to get the formatted current date
  function getFormattedDate() {
    // ge current date
    const currentDate = new Date();

    // Format current date to match the string type from input date
    const formattedDate = currentDate.toISOString().split('T')[0];

    return formattedDate;
  }

  const navigate = useNavigate();

  const handleChange = (e) => {
    const { id, value } = e.target;
    setBookingData((prev) => ({ ...prev, [id]: value }));
  };

  const handleAddDoctor = async (e) => {
    e.preventDefault();

    try {
      const rsp = await dispatch(createAppointment(bookingData));

      if (rsp.meta.requestStatus === 'fulfilled') {
        showAlert('Appointment booked successfully');
        navigate('/myappointments');
      }
    } catch (error) {
      console.log(error);
    }
  };

  return (
    <main className='bookDoctor d-flex flex-column '>
      <section className='d-flex'>
        <div
          className='back '
          onClick={() => {
            navigate(-1);
          }}
        >
          <BsCaretLeft />
        </div>
      </section>

      <section className=' d-flex flex-column align-items-center'>
        <h3 className='my-3'>
          {' '}
          Book Appointment with Doctor {bookingData.doctorname}
        </h3>

        <form onSubmit={handleAddDoctor} className='col-12 col-md-8 mx-auto'>
          <div className='my-3 col-12'>
            <label htmlFor='User Name'> User Name</label>
            <input
              type='text'
              id='username'
              placeholder='Enter user name'
              value={bookingData.username}
              className='form-control'
              onChange={handleChange}
              minLength={3}
              required
            />
          </div>
          <div className='my-3 col-12'>
            <label htmlFor='User ID'> User ID</label>
            <input
              type='text'
              id='user_id'
              placeholder='Enter user id'
              value={bookingData.user_id}
              className='form-control'
              onChange={handleChange}
              minLength={3}
              required
            />
          </div>
          <div className='my-3 col-12'>
            <label htmlFor='Doctor Name'> Doctor Name</label>
            <input
              type='text'
              id='doctorname'
              placeholder='Enter doctor name'
              value={bookingData.doctorname}
              className='form-control'
              onChange={handleChange}
              minLength={3}
              required
            />
          </div>
          <div className='my-3 col-12'>
            <label htmlFor='Doctor ID'> Doctor ID</label>
            <input
              type='text'
              id='doctor_id'
              placeholder='Enter doctor id'
              value={bookingData.doctor_id}
              className='form-control'
              onChange={handleChange}
              minLength={3}
              required
            />
          </div>

          <div className='my-3 col-12'>
            <label htmlFor='Date of appointment'> Date of appointment</label>
            <input
              type='date'
              id='date_of_appointment'
              value={bookingData.date_of_appointment}
              className='form-control'
              onChange={handleChange}
              min={getFormattedDate()}
              required
            />
          </div>
          <div className='mt-3'>
            <button className='main-btn col-12' type='submit'>
              {isLoading ? <Spinner /> : 'Book Now'}
            </button>
          </div>
        </form>
      </section>
    </main>
  );
};

export default BookDoctor;
