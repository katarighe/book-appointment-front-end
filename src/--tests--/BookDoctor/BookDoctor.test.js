// BookDoctor.test.js
import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import BookDoctor from '../../Pages/BookDoctor/BookDoctor';
import { createAppointment, selectAppointment } from '../../Redux/Features/appointmentSlice';
import { selectUserData } from '../../Redux/Features/userAuthSlice';
import { MemoryRouter } from 'react-router-dom';

// Mock Redux store
const mockStore = configureStore([]);

describe('BookDoctor Component', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      appointmentSlice: {
        allAppointments: [],
        isLoading: false,
        isError: '',
        appointmentDetails: {
          doctor_id: '123',
          name: 'Dr. Example',
        },
      },
      userAuthSlice: {
        authUser: {
          name: 'John Doe',
          image_url: 'https://example.com/avatar.jpg',
        },
      },
    });
  });

  test('renders BookDoctor component snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter>
        <BookDoctor />
        </MemoryRouter>
        
      </Provider>
    );

    // Use the Jest snapshot matcher to compare the rendered component with the snapshot
    expect(asFragment()).toMatchSnapshot();
  });
});
