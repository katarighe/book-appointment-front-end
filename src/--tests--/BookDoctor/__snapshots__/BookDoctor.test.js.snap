// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDoctor Component renders BookDoctor component snapshot 1`] = `
<DocumentFragment>
  <main
    class="bookDoctor d-flex flex-column "
  >
    <section
      class="d-flex"
    >
      <div
        class="back "
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 12.796V3.204L4.519 8 10 12.796zm-.659.753-5.48-4.796a1 1 0 0 1 0-1.506l5.48-4.796A1 1 0 0 1 11 3.204v9.592a1 1 0 0 1-1.659.753z"
          />
        </svg>
      </div>
    </section>
    <section
      class=" d-flex flex-column align-items-center"
    >
      <h3
        class="my-3"
      >
         Book Appointment with Doctor Dr. Example
      </h3>
      <form
        class="col-12 col-md-8 mx-auto"
      >
        <div
          class="my-3 col-12"
        >
          <label
            for="User Name"
          >
             User Name
          </label>
          <input
            class="form-control"
            id="username"
            minlength="3"
            placeholder="Enter user name"
            required=""
            type="text"
            value="John Doe"
          />
        </div>
        <div
          class="my-3 col-12"
        >
          <label
            for="User ID"
          >
             User ID
          </label>
          <input
            class="form-control"
            id="user_id"
            minlength="3"
            placeholder="Enter user id"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="my-3 col-12"
        >
          <label
            for="Doctor Name"
          >
             Doctor Name
          </label>
          <input
            class="form-control"
            id="doctorname"
            minlength="3"
            placeholder="Enter doctor name"
            required=""
            type="text"
            value="Dr. Example"
          />
        </div>
        <div
          class="my-3 col-12"
        >
          <label
            for="Doctor ID"
          >
             Doctor ID
          </label>
          <input
            class="form-control"
            id="doctor_id"
            minlength="3"
            placeholder="Enter doctor id"
            required=""
            type="text"
            value="123"
          />
        </div>
        <div
          class="my-3 col-12"
        >
          <label
            for="Date of appointment"
          >
             Date of appointment
          </label>
          <input
            class="form-control"
            id="date_of_appointment"
            required=""
            type="date"
            value=""
          />
        </div>
        <div
          class="mt-3"
        >
          <button
            class="main-btn col-12"
            type="submit"
          >
            Book Now
          </button>
        </div>
      </form>
    </section>
  </main>
</DocumentFragment>
`;
